pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java

```diff
@@ -388,6 +376,17 @@ private int getRestrictedMaxEntriesForConsumer(Consumer consumer, List<Entry> en
         return maxMessages;
     }
 
+    private boolean isEntryPendingAck(long ledgerId, long entryId) {
+        int size = consumerList.size();
+        for (int i = 0; i < size; i++) {
+            Consumer consumer = consumerList.get(i);
+            if (consumer != null && consumer.isPendingAck(ledgerId, entryId)) {
+                return true;
+            }
+        }
+        return false;
+    }
```

Comment:
It's super expensive if you have high traffic and many consumers, No?

Reply:
checking pending-ack is not expensive as it checks from the map. also, it's not being executed in every single  message but it gets checked if broker sees filtered messages and also only if message is not deleted. so, it won't be expensive and frequently executed.

regarding test cases, I had to change them because most of them were assuming consumers were stuck on unack messages on closed consumer and some of them are flaky as this test class is part of flaky test pipeline.


> I'm afraid the change will introduce new regressions even if it can fix some issues. I will leave change request here to make sure everything is clear on this PR before merge it.

Sure, let's make sure, there are no regression issues and change doesn't violate the user contract. this seems a fundamental fix to avoid all hack and stuck issue.  but you guys can verify and merge if you don't see any issue.

Reply:
If you assent with my approach to only fix https://github.com/apache/pulsar/issues/21656. I can also create a PR and we can keep this PR open to find a solution for the potential issues that you pointed out (simply adding removed consuner's unack messages into redelivery/replay list will not work because there are scenarios where key-shared dispatcher adds additionally filtered messages(due to max-message limit, or ordering) into replay list).

Reply:
sure, let's move forward with the best approach and let's make sure to fix to stuck dispatching as it's impacting many production systems right now and not giving positive experience about Pulsar.